### PR TITLE
Schema extension through SDL

### LINKED ISSUE #240: Schema extension through SDL
We need to add support for extending a schema through SDL. The reference implementation can be found here: https://github.com/graphql/graphql-js/pull/1314

### PR DESCRIPTION
**Pull request by crisu83**
Add support for schema extension through SDL (fixes #240).

### CODE REVIEW

File: src/Language/Node/SchemaDefinitionNode.php [hunk 1]
```diff
@@ -68,4 +68,20 @@ public function toAST(): array
             'loc'            => $this->getLocationAST(),
         ];
     }
+
+    /**
+     * @return NameNode|null
+     */
+    public function getName(): ?NameNode
+    {
+        // TODO: Implement getName() method.
```
Comment (Jalle19) [inline src/Language/Node/SchemaDefinitionNode.php]: Can we get these fixed before merging?

File: src/Language/Parser.php [hunk 2]
```diff
@@ -1307,6 +1310,42 @@ protected function lexTypeSystemExtension(): TypeExtensionNodeInterface
...
+    protected function lexSchemaExtension(): SchemaExtensionNode
+    {
+        $start = $this->lexer->getToken();
+
+        $this->expectKeyword(KeywordEnum::EXTEND);
+        $this->expectKeyword(KeywordEnum::SCHEMA);
+
+        $directives = $this->lexDirectives(true);
+
+        $parseFunction = function (): OperationTypeDefinitionNode {
```
Comment (Jalle19) [inline src/Language/Parser.php]: Does static return type-hinting really work for anonymous functions?

File: src/Schema/Extension/ExtensionContext.php [hunk 3]
```diff
@@ -59,14 +59,44 @@ public function isSchemaExtended(): bool
...
+    /**
+     * @return ObjectType[]
+     * @throws ExtensionException
+     * @throws InvariantException
+     */
+    public function getExtendedOperationTypes(): array
+    {
+        /** @noinspection PhpUnhandledExceptionInspection */
+        $operationTypes = [
+            'query' => $this->getExtendedQueryType(),
```
Comment (Jalle19) [inline src/Schema/Extension/ExtensionContext.php]: Vertical alignment

File: src/Schema/Extension/ExtensionContext.php [hunk 4]
```diff
@@ -106,13 +136,17 @@ public function getExtendedSubscriptionType(): ?TypeInterface
      */
     public function getExtendedTypes(): array
     {
+        $typeMap       = $this->info->getSchema()->getTypeMap();
```
Comment (Jalle19) [inline src/Schema/Extension/ExtensionContext.php]: Vertical alignment

File: tests/Functional/Schema/ExtensionTest.php [hunk 5]
```diff
@@ -571,6 +580,193 @@ public function testDoesNotAllowExtendingANonObjectScalarType()
         '));
     }
 
+    // can add additional root operation types
```
Comment (Jalle19) [inline tests/Functional/Schema/ExtensionTest.php]: I'd drop these comments

File: src/Language/Parser.php (hunk 2 above)
Comment (crisu83) [reply src/Language/Parser.php]: Yes they do. :)

File: src/Schema/Extension/ExtensionContext.php (hunk 3 above)
Comment (crisu83) [reply src/Schema/Extension/ExtensionContext.php]: Fixed.

File: tests/Functional/Schema/ExtensionTest.php (hunk 5 above)
Comment (crisu83) [reply tests/Functional/Schema/ExtensionTest.php]: I'll keep this one, it's from the reference implementation.

**Pull request merged**